### DISCUSSION TITLE
The app.shutdown() takes about 30 seconds to shut down?

### DISCUSSION THREAD

**Discussion by v479038280** [zauberzeug/nicegui#3401] [General]
The app.shutdown() takes about 30 seconds to shut down,is it normal?
The program contains only one line code ,with no additional logi.
```py
from nicegui import app, ui

ui.button('shutdown', on_click=app.shutdown)

ui.run()
```

**Comment by v479038280** [comment]
Here is another version ,it works fine.
```py
from nicegui import app, ui
from datetime import datetime

def shutdown_():
    print(f'start shutdown:  {datetime.now()}')
    app.shutdown()
ui.button('shutdown', on_click=shutdown_)
app.on_shutdown(lambda: print(f'shutdown:  {datetime.now()}'))

ui.run()
```

The result:
```
start shutdown:  2024-07-27 12:52:03.628222
shutdown:  2024-07-27 12:52:03.779497
```

**Reply by v479038280** [reply]
This one also works fine.
```py
from nicegui import app, ui
from datetime import datetime

ui.button('shutdown', on_click=lambda:(print(datetime.now(),app.shutdown())))
app.on_shutdown(lambda: print(f'shutdown:  {datetime.now()}'))

ui.run()
```

here is the result:
```
2024-07-27 12:50:14.455136 None
shutdown:  2024-07-27 12:50:14.653358
```

**Comment by python-and-novella** [comment]
It  works well,need more details of your environment.

**Reply by v479038280** [reply]
I'm using python 3.11.5 on windows10 professional with the latest nicegui 1.4.30

**Reply by python-and-novella** [reply]
run it directly? not debug in ide?

**Reply by v479038280** [reply]
yes,I run it directly

**Reply by v479038280** [reply]
```py
from nicegui import app, ui
from datetime import datetime

ui.button('shutdown', on_click=lambda:app.shutdown())
app.on_shutdown(lambda: print(f'shutdown:  {datetime.now()}'))

ui.run(reload=False)
```
This will also shutdown immediately

**Comment by v479038280** [comment]
> The app.shutdown() takes about 30 seconds to shut down,is it normal? The program contains only one line code ,with no additional logi.
> 
> ```py
> from nicegui import app, ui
> 
> ui.button('shutdown', on_click=app.shutdown)
> 
> ui.run()
> ```

```py
from nicegui import app, ui
from datetime import datetime

ui.button('shutdown', on_click=lambda:app.shutdown())
app.on_shutdown(lambda: print(f'shutdown:  {datetime.now()}'))

ui.run()
```
This is the same as above

**Comment by v479038280** [comment]
I test it in Powershell ,it works fine. But when I run it in  Vscode ,it has such problem, the terminal with Vscode is actually Powershell

**Reply by python-and-novella** [reply]
It is a bug of windows terminal.
See it:
Launching in Russian layout causes Ctrl+C and other signal chars to not work · Issue #17197 · microsoft/terminal
https://github.com/microsoft/terminal/issues/17197
Console ReadFile canceled by Ctrl+C fails to set ERROR_OPERATION_ABORTED · Issue #334 · microsoft/terminal
https://github.com/microsoft/terminal/issues/334
Ctrl+C in terminal broken on windows · Issue #1972 · encode/uvicorn
https://github.com/encode/uvicorn/issues/1972

**Reply by python-and-novella** [reply]
I try it in native `cmd` and native `powershell`, it's ok.
It takes about 25s to shutdown in `Windows Terminal` when  `Windows Terminal` loses focus.
![image](https://github.com/user-attachments/assets/c9cd16cf-26fd-4b34-a24e-a44c5aff00c8)

**Comment by python-and-novella** [comment]
```python3
from nicegui import app, ui

ui.button('shutdown1', on_click=lambda :(app.shutdown(), print('\r', end='',sep='')))

ui.run()
```

**Reply by v479038280** [reply]
We could use it in such way

**Reply by python-and-novella** [reply]
@falkoschindler 
Should we change `shutdown` in `app.py`?
Change
```python3
os.kill(os.getppid(), getattr(signal, 'CTRL_C_EVENT' if platform.system() == 'Windows' else 'SIGINT'))
```
to
```python3
os.kill(os.getppid(), getattr(signal, 'CTRL_C_EVENT' if platform.system() == 'Windows' else 'SIGINT'))
print('\r', end='',sep='')
```
It only happened when terminal loses focus during shutdowning before timeout. But when terminal gets focus,it will react immediately.

**Reply by falkoschindler** [reply]
@python-and-fiction When implementing PR #3320, we decided not to add `\r` because it felt quite hacky and it is ultimately a bug in VSCode - as far as we understand. So I recommend adding a wrapper function in user code if running in VSCode is too much of a hassle.
